algo/src/main/java/org/neo4j/graphalgo/impl/AllShortestPaths.java

```diff
@@ -0,0 +1,142 @@
+        this.graph = graph;
+        this.nodeCount = graph.nodeCount();
+        this.executorService = executorService;
+        if (concurrency < 1) {
+            throw new IllegalArgumentException("concurrency must be >0");
+        }
+        this.concurrency = concurrency;
+        this.counter = new AtomicInteger();
+        this.resultQueue = new LinkedBlockingDeque<>(); // TODO limit size?
+    }
+
+
+    public Stream<Result> resultStream() {
+
+        counter.set(0);
+        running = true;
+
+        for (int i = 0; i < concurrency; i++) {
+            executorService.submit(new ShortestPathTask());
+        }
+
+        long end = (long) nodeCount * nodeCount;
+
+        return LongStream.range(0, end)
+                .onClose(() -> running = false)
+                .mapToObj(i -> {
+                    try {
+                        return resultQueue.take();
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+    }
+
+    private class ShortestPathTask implements Runnable {
+
+        private final IntMinPriorityQueue queue;
+        private final double[] distance;
+
+        private ShortestPathTask() {
+            distance = new double[nodeCount];
+            queue = new IntMinPriorityQueue();
+        }
+
+        @Override
+        public void run() {
+            int startNode;
+            while ((startNode = counter.getAndIncrement()) < nodeCount) {
+                compute(startNode);
+                for (int i = 0; i < nodeCount; i++) {
+                    final Result result = new Result(
+                            graph.toOriginalNodeId(startNode),
+                            graph.toOriginalNodeId(i),
+                            distance[i]);
+
+                    try {
+                        resultQueue.put(result);
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            }
+        }
+
+        public void compute(int startNode) {
+            Arrays.fill(distance, Double.POSITIVE_INFINITY);
+            distance[startNode] = 0d;
+            queue.add(startNode, 0d);
+
+            while (!queue.isEmpty()) {
+                final int node = queue.pop();
+                final double sourceDistance = distance[node];
+                // scan relationships
+                graph.forEachRelationship(
+                        node,
+                        Direction.OUTGOING,
+                        (source, target, relId, weight) -> {
+                            // relax
+                            final double targetDistance = weight + sourceDistance;
+                            if (targetDistance < distance[target]) {
+                                distance[target] = targetDistance;
+                                queue.add(target, targetDistance);
```

Comment:
this will return false if it was not added, and not block,
consider using offer() instead